src/shogun/util/visitors/FilterVisitor.h

```diff
@@ -0,0 +1,168 @@
+#ifndef SHOGUN_FILTERVISITOR_H
+#define SHOGUN_FILTERVISITOR_H
+
+#include <shogun/base/SGObject.h>
+
+#include <functional>
+
+namespace shogun
+{
+
+	template <typename T>
+	class FilterVisitor : public AnyVisitor
+	{
+
+	public:
+		FilterVisitor(std::function<void(const std::string&, T*)> operation)
+		    : AnyVisitor(), m_operation(operation)
+		{
+		}
+
+		template <typename U>
+		void on_impl(U* v)
+		{
+			if constexpr (std::is_pointer<U>::value)
+				if (!*v)
+					return;
+
+			if constexpr (std::is_same<T, U>::value)
+				m_operation(m_name, v);
+
+			if constexpr (std::is_base_of<
+			                  typename std::remove_pointer<U>::type,
+			                  typename std::remove_pointer<T>::type>::value)
+			{
+				auto obj = dynamic_cast<T>(*v);
```

Comment:
isn't it T*?

Reply:
T* would be SGObject** or Kernel**. I'm not sure we can cast to a pointer pointer.

Reply:
ah ok, sorry!